artifactory/services/go/publish.go

```diff
@@ -1,28 +1,117 @@
 
-type PublishGoPackage interface {
-	isCompatible(artifactoryVersion string) bool
-	PublishPackage(params GoParams, client *jfroghttpclient.JfrogHttpClient, ArtDetails auth.ServiceDetails) (*utils.OperationSummary, error)
+// Support for Artifactory 6.6.1 and above API
+type GoPublishCommand struct {
+	artifactoryVersion string
+	clientDetails      httputils.HttpClientDetails
+	client             *jfroghttpclient.JfrogHttpClient
 }
 
-func register(publishApi PublishGoPackage) {
-	publishers = append(publishers, publishApi)
+func (pwa *GoPublishCommand) isCompatible(artifactoryVersion string) bool {
+	propertiesApi := "6.6.1"
+	version := version.NewVersion(artifactoryVersion)
+	pwa.artifactoryVersion = artifactoryVersion
+	return version.AtLeast(propertiesApi)
 }
 
-// Returns the compatible publisher to Artifactory
-func GetCompatiblePublisher(artifactoryVersion string) PublishGoPackage {
-	for _, publisher := range publishers {
-		if publisher.isCompatible(artifactoryVersion) {
-			return publisher
+func (pwa *GoPublishCommand) PublishPackage(params GoParams, client *jfroghttpclient.JfrogHttpClient, ArtDetails auth.ServiceDetails) (*utils.OperationSummary, error) {
+	url, err := utils.BuildArtifactoryUrl(ArtDetails.GetUrl(), "api/go/"+params.GetTargetRepo(), make(map[string]string))
+	if err != nil {
+		return nil, err
+	}
+	pwa.clientDetails = ArtDetails.CreateHttpClientDetails()
+	pwa.client = client
+	moduleId := strings.Split(params.GetModuleId(), ":")
+	totalSucceed, totalFailed := 0, 0
+	var filesDetails []clientutils.FileTransferDetails
+	// Upload zip file
+	success, failed, err := uploadFile(params, params.ZipPath, moduleId[0], ".zip", url, &filesDetails, pwa)
+	if err != nil {
+		return nil, err
+	}
+	totalSucceed, totalFailed = totalSucceed+success, totalFailed+failed
+	// Upload mod file
+	success, failed, err = uploadFile(params, params.ModPath, moduleId[0], ".mod", url, &filesDetails, pwa)
+	if err != nil {
+		return nil, err
+	}
+	totalSucceed, totalFailed = totalSucceed+success, totalFailed+failed
+	if version.NewVersion(pwa.artifactoryVersion).AtLeast(ArtifactoryMinSupportedVersionForInfoFile) && params.GetInfoPath() != "" {
+		// Upload info file. This is supported from Artifactory version 6.10.0 and above
+		success, failed, err = uploadFile(params, params.InfoPath, moduleId[0], ".info", url, &filesDetails, pwa)
+		totalSucceed, totalFailed = totalSucceed+success, totalFailed+failed
+		if err != nil {
+			return nil, err
 		}
 	}
-	return nil
+	tempFile, err := clientutils.SaveFileTransferDetailsInTempFile(&filesDetails)
+	if err != nil {
+		return nil, err
+	}
+	return &utils.OperationSummary{TotalSucceeded: totalSucceed, TotalFailed: totalFailed, TransferDetailsReader: content.NewContentReader(tempFile, "files")}, nil
+}
+
+func uploadFile(params GoParams, filePath string, moduleId, ext, url string, filesDetails *[]clientutils.FileTransferDetails, pwa *GoPublishCommand) (success, failed int, err error) {
```

Comment:
For symmetry reasons, let's add this function to the ```GoPublishCommand``` struct, like the rest fo the methods in this file.